### PR TITLE
Reworking Logger

### PR DESCRIPTION
**Pull request by KevinPlus3**
Added new private method for setting up logger. Refactored code and made some minor improvements. Also addresses issues #60 to #63.

---
This change is [Reviewable](https://reviewable.io/reviews/plus3it/watchmaker/69)

### CODE REVIEW

File: src/watchmaker/__init__.py [hunk 1]
```diff
@@ -108,30 +135,35 @@ def _get_config_data(self):
...
-                      'work!\nPlease make sure your config is available.')
+                logging.critical(
+                    'The URL used to get the user config.yaml file did not '
+                    'work!\nPlease make sure your config is available.'
+                )
                 sys.exit(1)
 
         if self.config_path and not os.path.exists(self.config_path):
-            self.logger.warning(
+            logging.warning(
```
Comment (lorengordon) [inline src/watchmaker/__init__.py]: I feel like this should probably be a critical error, log and exit. As is, getting here means the user passed in a param, the param is not validating properly, but we are proceeding anyway with a config different than they intended. That seems counter to the principle of least surprise.

File: src/watchmaker/__init__.py [hunk 2]
```diff
@@ -108,30 +135,35 @@ def _get_config_data(self):
...
             data = f.read()
 
         if data:
             self.config = yaml.load(data)
         else:
-            logging.debug('No data to load.')
+            logging.critical(
+                'Unable to load the data of the default or'
+                ' the user supplied config.'
+            )
```
Comment (lorengordon) [inline src/watchmaker/__init__.py]: I think the only way to get into this branch is if the file is empty. This should probably exit if there is no config data, right? Does `logging.critical()` automatically exit?

Comment (KevinPlus3) [inline src/watchmaker/__init__.py]: Yeah. I was asking Adam about that earlier when I was like shouldn't this be critical.  `logging.critical` does not automatically exit (I don't think).  But I think we may need to do some refactoring in terms of properly exiting.  I can add a sys.exit(1) as well for now.

File: src/watchmaker/__init__.py [hunk 3]
```diff
@@ -51,42 +47,73 @@ def __init__(self, arguments):
...
+        elif log_file:
+            if os.path.isdir(log_file):
+                log_msg = '{0} is a directory and not a file.'.format(log_file)
+                log_type = 'error'
+            else:
+                log_filename = log_file
+                log_msg = 'Start time: {0}'.format(datetime.datetime.now())
+                log_type = 'info'
+        elif not log_dir and not log_file:
+            log_msg = 'Stream logger is not enabled!'
```
Comment (lorengordon) [inline src/watchmaker/__init__.py]: I don't think users are going to know what a stream logger is. (I still don't....) What does this really mean, from the user's perspective?

File: src/watchmaker/__init__.py [hunk 4]
```diff
@@ -108,30 +135,35 @@ def _get_config_data(self):
                     shutil.copyfileobj(response, outfile)
                 self.config_path = 'config.yaml'
             except urllib.error.URLError:
-                print('The URL used to get the user config.yaml file did not '
-                      'work!\nPlease make sure your config is available.')
+                logging.critical(
+                    'The URL used to get the user config.yaml file did not '
+                    'work!\nPlease make sure your config is available.'
```
Comment (lorengordon) [inline src/watchmaker/__init__.py]: Thoughts on moving away from multiline messages, now that we've moved to the logging facility? I think they make the log file ugly. If we need multiple lines, we maybe should use multiple `logging()` calls.

File: src/watchmaker/cli.py [hunk 5]
```diff
@@ -15,15 +15,16 @@ def main():
...
-    parser.add_argument('--logger', dest='logger', action='store_true',
-                        help='Use stream logger for debugging.')
-    parser.add_argument('--log-path', dest='log_path', default=None,
-                        help='Path to the logfile for stream logging.')
     parser.add_argument('--saltstates', dest='saltstates', default=None,
                         help=(
                             'Define the saltstates to use.  Must be None, '
                             'Highstate, or comma-seperated-string'
                         ))
+    group = parser.add_mutually_exclusive_group(required=False)
```
Comment (lorengordon) [inline src/watchmaker/cli.py]: Nice, was wondering how we were going to do that, turns out it's just part of argparse. Looking down the road, maybe name this `log_group`, so if we have another set of mutually exclusive params the naming continues to make sense?

**Comment by lorengordon** [comment]
A few comments, but looks good overall.

File: src/watchmaker/__init__.py (hunk 3 above)
Comment (KevinPlus3) [inline src/watchmaker/__init__.py]: Probably nothing to a user at the top level.  I'll shorten to just logger.

File: src/watchmaker/__init__.py (hunk 4 above)
Comment (KevinPlus3) [inline src/watchmaker/__init__.py]: I agree with you on this front.  If we want it to be similar to typical logs, then most if not all actually keep info on one line per log.

File: src/watchmaker/cli.py (hunk 5 above)
Comment (KevinPlus3) [inline src/watchmaker/cli.py]: :thumbs_up: Will do.

**Comment by lorengordon** [comment]
Reviewed 3 of 5 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/plus3it/watchmaker/69)*

**Pull request merged**